st2api/st2api/controllers/resource.py

```diff
@@ -368,6 +400,61 @@ def _validate_exclude_fields(self, exclude_fields):
         return exclude_fields
 
 
+class BaseResourceIsolationControllerMixin(object):
+    """
+    Base API controller which isolates resources for users. Users can only see their own resources.
+
+    Exceptions include admin and system user which can view all the resources (also for other
+    users).
+    """
+
+    def resources_model_filter(self, model, instances, requester_user=None, offset=0, eop=0,
+                              **from_model_kwargs):
+        # RBAC or permission isolation is disabled, bail out
+        if not (cfg.CONF.rbac.enable and cfg.CONF.rbac.permission_isolation):
+            result = super(BaseResourceIsolationControllerMixin, self).resources_model_filter(
+                model=model, instances=instances, requester_user=requester_user,
+                offset=offset, eop=eop, **from_model_kwargs)
+
+            return result
+
+        result = []
+        for instance in instances[offset:eop]:
```

Comment:
Is this going to cause bottleneck in performance?

Reply:
Any kind of filtering will definitely decrease the performance, but I think it should be fine since we only do it for non-system (non-stanley) users.

Reply:
And to clarify further - we wanted something like that from the beginning but one of the reasons why I avoided implementing it (in addition to customers not requesting it) was that implementing it in a performance efficient manner would be complex and require caching, etc.

In this case, it's opt-in so it's not a big deal and we can sort out the performance (introduce caching, etc). later, if needed.